### ISSUE TITLE
SparsityPattern::Build::operator() optimization

### ISSUE THREAD

**Issue by permcody** [libMesh/libmesh#586]
While profiling Rattlesnake, we found a rather expensive operation in `SparsityPattern::Build::operator()`

When running a simulation with ~1300 variables the "if" statement in the following loop was taking about 50% of the time in this routine which itself was taking around 30% of the total setup time (several minutes). The issue is with using a dense matrix that is mostly zero for this operation. It should be replaced with a different data structure.

The problem in question can be found here:
`rattlesnake/examples/C5G7/diffusion_c5g7.i`

```
            for (unsigned int vj=0; vj<n_var; vj++)
              if ((*dof_coupling)(vi,vj)) // If vi couples to vj
                {
                  // Find element dofs for variable vj, note that
                  // if vi==vj we already have the dofs.
                  if (vi != vj)
```

**Comment by roystgnr** [comment]
@benkirk I could swear you used the CouplingMatrix to optimize FIN-S, but I can't find it in the code.  Am I just misremembering?

It looks like we'll be able to be backwards-compatible the existing API with a sparse matrix structure.  Actually *using* the existing API won't be efficient, but it'll be an O(N_vars^2) inefficiency, not an O(N_vars^2*N_elem) inefficiency.  Even so we'll want a new API (appending sparse indices to a row, not directly modifying i,j) for newer code.

**Comment by benkirk** [comment]
I don't, all the variables are coupled there. But I do rely on the variable groups, so that might be the confusion.

-Ben

On Jun 3, 2015, at 3:35 PM, roystgnr <notifications@github.com<mailto:notifications@github.com>> wrote:


@benkirk<https://github.com/benkirk> I could swear you used the CouplingMatrix to optimize FIN-S, but I can't find it in the code. Am I just misremembering?

It looks like we'll be able to be backwards-compatible the existing API with a sparse matrix structure. Actually using the existing API won't be efficient, but it'll be an O(N_vars^2) inefficiency, not an O(N_vars^2*N_elem) inefficiency. Even so we'll want a new API (appending sparse indices to a row, not directly modifying i,j) for newer code.

**Comment by roystgnr** [comment]
@dknez, I see some CouplingMatrix usage in the Reduced Basis code.  Is there any in your private app codes?

I'm trying to figure out how safe it is for me to break the API.  "operator() returns a reference" is what we've got now, but that entails a bunch of wasteful storage.  Returning a proxy would work for user apps that use that method implicitly (the way RB and MOOSE do), but would break for anyone who actually tried to assign a unsigned char&.  Removing the non-const operator() entirely would be easiest but would require everyone who sets a CouplingMatrix to switch to a new API.

**Comment by dknez** [comment]
Yes, CouplingMatrix is used in the Reduced Basis code --- in RBEIMConstruction to be specific --- and RBEIMConstruction is used by our app codes.

The examples reduced_basis_ex4 and reduced_basis_ex_6 use RBEIMConstruction, so as long as those examples still work fine, then everything should be OK from my end. But it'd be good to do a before and after comparison for those examples to check that nothing changes (and to check there is no slow-down, since the coupling matrix is used for efficiency reasons).

**Comment by roystgnr** [comment]
Fixed by #589
